File: bindings/java/src/main/java/com/aster/ffi/IrohException.java
package com.aster.ffi;

/** Top-level exception for all FFI-level errors. */
public class IrohException extends RuntimeException {
  private final IrohStatus status;
  private final int errorCode;

  public IrohException(IrohStatus status, String message) {
    super(message);
    this.status = status;
    this.errorCode = -1;
  }

  public IrohException(IrohStatus status, int errorCode, String message) {
    super(message);
    this.status = status;
    this.errorCode = errorCode;
  }

  public IrohException(String message) {
    super(message);
    this.status = IrohStatus.INTERNAL;
    this.errorCode = -1;
  }

  public IrohStatus status() {
    return status;
  }

  public int errorCode() {
    return errorCode;
  }
}

File: bindings/dotnet/src/Aster/IrohException.cs
using System;

namespace Aster;

/// <summary>
/// Thrown when an iroh FFI call returns a non-zero status code.
/// </summary>
public class IrohException : Exception
{
    public Status Status { get; }

    public IrohException(Status status, string message) : base(message)
    {
        Status = status;
    }

    public IrohException(string message) : base(message)
    {
        Status = Status.Internal;
    }

    public static IrohException FromStatus(int code, string context)
    {
        var status = (Status)code;
        string name = Native.iroh_status_name((uint)code) != IntPtr.Zero
            ? Marshal.PtrToStringAnsi(Native.iroh_status_name((uint)code)) ?? status.ToString()
            : status.ToString();
        return new IrohException(status, $"{context}: {name} ({code})");
    }
}
